### ISSUE TITLE
required validation bug

### ISSUE THREAD

**Issue by javiercbk** [Automattic/mongoose#7249]
**Do you want to request a *feature* or report a *bug*?**
Bug
**What is the current behavior?**
When setting a model property as required inside an array, the validation path goes out of sync when using splice (thus changing the index).
**If the current behavior is a bug, please provide the steps to reproduce.**
```js
const mongoose = require('mongoose');

const childSchema = mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Child name is required'],
  }
});

const kittySchema = new mongoose.Schema({
  name: String,
  children: [childSchema],
});

const Kitten = mongoose.model('Kitten', kittySchema);

const kitten = new Kitten({ name: 'Kitten 1', children: [
  { name: 'Child 1'},
  { name: 'Child 2'},
]});

// remember to change the username, pass and db
mongoose.connect('mongodb://user:pass@localhost/db');
const db = mongoose.connection;
db.on('error', console.error.bind(console, 'connection error:'));
db.once('open', function() {
  return kitten.save().then(() => {
    return Kitten.findOne({ _id: kitten._id });
  }).then((k) => {
    // this will trigger the validation, if this line is commented the bug will not manifest
    k.children[1].name = 'Child 3';
    // this will screw up the validation
    k.children.splice(0, 1);
    k.markModified('children');
    // uncomment the following lines and it will work
    // const aux = k.children;
    // k.children = [];
    // k.children = aux;
    return k.save();
  }).then(() => {
    console.log('success');
  }).catch((err) => {
    console.log(JSON.stringify(err));
  //   {
  //     "errors": {
  //         "children.1.name": {
  //             "message": "Child name is required",
  //             "name": "ValidatorError",
  //             "properties": {
  //                 "message": "Child name is required",
  //                 "type": "required",
  //                 "path": "name"
  //             },
  //             "kind": "required",
  //             "path": "name"
  //         }
  //     },
  //     "_message": "Kitten validation failed",
  //     "message": "Kitten validation failed: children.1.name: Child name is required",
  //     "name": "ValidationError"
  // }
  });
});
```

**What is the expected behavior?**
The validation path should be updated when splice is called over an array. I think the problem is that the document calculates the validation paths and do not update those path when splice is called.
**Please mention your node.js, mongoose and MongoDB version.**
node: v10.6.0
mongoose: 5.3.12
mongodb: 3.4.13

**Comment by vkarpov15** [comment]
Thanks for reporting, will fix ASAP :+1:

**Comment by YannisMarios** [comment]
Are you sure this is fixed? Because I get the same error on v5.4.3 even though I provide required field.

activity.js
```
const ActivitySchema = new Schema({
  name: {
   type: String,
   required: true
  }
});
export default mongoose.model('activity', ActivitySchema );
```

taxOffice.js
```
const TaxOfficeSchema = new Schema({
  name: {
   type: String,
   required: true
  }
});
export default mongoose.model('taxOffice', TaxOfficeSchema);
```
company.js
```
const companySchema = new Shema({
  activity: {
    type: Schema.Types.ObjectId,
    ref: 'activity',
    required: true
  },
  taxOffice: {
    type: Schema.Types.ObjectId,
    ref: 'taxOffice',
    required: true
  }
});

```

then in my apollo-graphql resolver for saving company
(data below contains activity _id and taxOffice _id
```
async saveCompany({data}) ==> {
  const 
}
```

**Comment by YannisMarios** [comment]
Anyone else facing a similar issue?
